### DISCUSSION TITLE
Tailwind @apply not working

### DISCUSSION THREAD

**Discussion by danhodgins** [tailwindlabs/tailwindcss#1889] [Help] [answered]
Hello, first let me say I'm having a blast with Tailwind. Really enjoying it so far, so thanks for all the hard work that has been done already. 

After following the standard Tailwind installation procedure I'm not able to get @apply to work to extract component classes. 

---

Here is my css/tailwind.css:
```
@tailwind base;
@tailwind components;
.btn {
	@apply bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded;
}
@tailwind utilities;  
```
---

And the HTML: 

```
<button class="btn">My Button</button>
<a href="#" class="btn">My Button</button>

```
---
Stylesheet links in HTML
```
<link rel="stylesheet" href="build/tailwind.css">
     
         <link
            rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/material-design-icons/3.0.1/iconfont/material-icons.min.css"
            integrity="sha256-x8PYmLKD83R9T/sYmJn1j3is/chhJdySyhet/JuHnfY=" crossorigin="anonymous" />
         <link
            href="https://fonts.googleapis.com/css2?family=Rubik:wght@300;400;500;700;900&display=swap" rel="stylesheet"/>
```

The result: 
The styles are not applied to the .btn class. 

---
I have followed to installation instructions and tried to find other threads of other people facing this problem, but couldn't find anything, hence the question here. 

Has @apply been deprecated perhaps in favour of using template partials? 

Well done on an enjoyable utility first CSS framework! Again, I'm loving it so far : ) 

Dan

**Comment by estevanmaito** [comment]
From the docs: https://tailwindcss.com/docs/functions-and-directives/#apply

> It's important to understand that @apply will not work for inlining pseudo-class or responsive variants of another utility. Instead, apply the plain version of that utility into the appropriate pseudo-selector or a new media query:

```css
/* Won't work: */
.btn {
  @apply block bg-red-500;
  @apply hover:bg-blue-500;
  @apply md:inline-block;
}

/* Do this instead: */
.btn {
  @apply block bg-red-500;
}
.btn:hover {
  @apply bg-blue-500;
}
@screen md {
  .btn {
    @apply inline-block;
  }
}
```

This recent issue points to other issues related (see first comment): https://github.com/tailwindcss/tailwindcss/issues/1888

**Comment by danhodgins** [comment]
Good point regarding pseudo classes @estevanmaito. 

I have changed the code in: css/tailwind.css to the following to eliminate any pseudo class issue (just for testing purposes): 

```
@tailwind base;
@tailwind components;
@tailwind utilities;

.btn {
	@apply bg-blue-500;
}

.btn {
	@apply text-white; 
}

.btn {
	@apply uppercase; 
}

.btn {
	@apply font-bold; 
}

.btn {
	@apply py-2; 
}

.btn {
	@apply  px-4; 
}

.btn {
	@apply rounded; 
}
```

This does not work. 

According to the Tailwind docs I should be able to do this and have it work: 

```
.btn {
	@apply bg-blue-500 text-white font-bold py-2 px-4 rounded;
}
```
Status: @apply still does not work for me no matter what I try.

Is there anything anywhere that might be case sensitive, indent sensitive, etc?

**Reply by estevanmaito** [reply]
Could you share your config files? Tailwind, PostCSS (if any)...

In your `build/tailwind.css`, is the `.btn` the only missing class?

**Reply by danhodgins** [reply]
Yes, in my build/tailwind.css, .btn is the only missing class. Should I add .btn there? 

I installed via NPM (using Adam's YT tutorial). 

Right now I'm trying to add the @apply directive in css/tailwind.css, not in /build/tailwind.css. 

Note: I am trying to use @apply by building a static HTML site. I am viewing the HTML in Google Chrome using file:/// on my Mac.  Not sure if that would somehow disable @apply. 

Here are the files: 

**public/build/tailwind.css**
This file is unchanged from the original installation.

**css/tailwind.css**
```
@tailwind base;
@tailwind components;

.btn {
	@apply bg-blue-500 text-white uppercase font-bold px-4 rounded; 
}

@tailwind utilities;
```
---
**package.json**

```
{
  "name": "DB_2020",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "build": "postcss css/tailwind.css -o public/build/tailwind.css"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "autoprefixer": "^9.8.0",
    "postcss-cli": "^7.1.1",
    "tailwindcss": "^1.4.6"
  }
}
```
---
**postcss.config.js**

module.exports = {
	
	plugins: [
		require('tailwindcss'),
		require('autoprefixer'),
		
	]
}
----
**tailwind.config.js**

// tailwind.config.js
module.exports = {
  theme: {},
  variants: {},
  plugins: [],
}

---
**Stylesheet link on index.html**
```
<link rel="stylesheet" href="build/tailwind.css">
  
```
---
**Button test on index.html**
```
<button class="btn">My Button</button>
<a href="#" class="btn">My Button</button>
```

**Comment by danhodgins** [comment]
Thoughts @estevanmaito?

**Reply by estevanmaito** [reply]
I just tried it locally, and couldn't reproduce the problem. I get the same output with PostCSS and just Tailwind, production or development. In every case, `btn` is correctly compiled.

Is the `index.html` you're viewing, inside `public` folder?

That's what I can guess just by looking at what you provided. It should work. The only way I could help you would be by looking at a reproducible case/repo.

**Comment by danhodgins** [comment]
Here is a Github repo with my code @estevanmaito: 

(https://github.com/danhodgins/tailwind_test)

**Reply by estevanmaito** [reply]
no link 😅

**Reply by danhodgins** [reply]
Just added.

**Reply by estevanmaito** [reply]
I cloned your repo. Ran `npm install`, `npm run build` and `btn` is inside `public/build/tailwind.css`

I suggest you to delete your current local repo and start again, because there is apparently some cache or file issue that is not reproducible with a new installation.

*also, the root `build` is misleading and I think you could delete it.

**Reply by danhodgins** [reply]
OK, will try that right now.

**Comment by danhodgins** [comment]
Updated Github repo:

(https://github.com/danhodgins/tailwind_test7)

**Reply by estevanmaito** [reply]
Your config files point to places that doesn't exist, then your HTML look for files that doesn't exist...

If you just follow this 7 min screencast (skip the first minute), you'll end up with a project that will work 100%: https://tailwindcss.com/course/setting-up-tailwind-and-postcss/

**Reply by danhodgins** [reply]
OK, thank you for your help Estevan. Much appreciated. -Dan

**Comment by danhodgins** [comment, marked as answer]
Created a Rails 5 project with Tailwind using Webpacker, and the @apply directive is working properly. 

Who knows what was wrong before whe testing on my local machine, but it wasn't Tailwind itself : )

Big thanks to @estevanmaito for taking the time to troubleshoot.

**Comment by Besuf** [comment]
The issue is you should run the postcss command with -i. I was faced with similar issue and I was using the following script.
`"scripts": {
    "build-css": "npx tailwindcss-cli@latest build ./src/styles.css -o ./public/styles.css"
  },`

then I added -i in my command as follows

`"scripts": {
    "build-css": "npx tailwindcss-cli@latest build -i ./src/styles.css -o ./public/styles.css"
  },`

and everything worked. It has got to do with running without -i is deprecated.

**Reply by Fortrain** [reply]
Thanks a lot @Besuf. I was also facing the same problem for two days and didn't know what was wrong. Things are working properly now.

**Reply by TenzinJam** [reply]
Thank you. I was also having trouble with it. The console kept telling me to include the -i in my script, but I put it off. It worked as soon as I did so. Thank you, @Besuf !!

**Reply by salma-py97** [reply]
This is it. Thank you!!

**Reply by simonswiss** [reply]
Tailwind CSS since v2 comes with the CLI integrated in the `tailwindcss` package, and it's recommended to use that instead of the `tailwindcss-cli` package:

https://tailwindcss.com/docs/installation#using-tailwind-cli

It comes with support for the new `JIT` engine, a `--watch` flag to update on changes, and is really, _really_ fast 💨

**Reply by vaibhavsanil** [reply]
Yes !!! This worked for me Thanks

**Reply by dev-sonen** [reply]
thank you so much!

**Reply by bajahranks** [reply]
This was indeed the right approach, however @simonswiss is correct. Same ice-cream but with sprinkles.

**Comment by Filefabrik** [comment]
Sometimes it is good to remove the node_modules directory completely and run "npm install"

**Comment by hoseinsedaqat** [comment]
Thanks @Besuf  This way work it ❤

**Comment by yellow1912** [comment]
It should be noted that, according to the document of the latest version, this will work:

```
.btn-blue {
  @apply bg-blue-500 hover:bg-blue-700 text-white;
}
```

**Comment by rigeltoth** [comment]
Hey guys!
this worked for me: https://github.com/tailwindlabs/tailwindcss/discussions/5258

**Reply by navein97** [reply]
This worked for me, thanks!!!!!!!!!!

**Comment by SaymanRabbi** [comment]
Same Issue.
Does anyone use @apply in next.js

**Comment by ompandey1996** [comment]
learning tailwind not useing with any framework 
i have html css and config.js file  
and install  its dependencies  npm install  .... D post ..prefix  
       npx tail.... init 
please help

<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <script src="https://tailwindcss.com"></script>
</head>

<body>
    <button class="btn-primary">
        Save changes
      </button>
</body>

</html>   

css 
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .btn-primary {
    @apply py-2 px-4 bg-blue-500 text-white font-semibold rounded-lg shadow-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-400 focus:ring-opacity-75;
  }
}
